File: 39-combination-sum/combination-sum.go
func combinationSum(candidates []int, target int) [][]int {
	var output [][]int
	N := len(candidates)

	var findCombination func(idx int, currPath []int, currTotal int)
	findCombination = func(idx int, currPath []int, currTotal int) {
		if currTotal == target {
			copyPath := make([]int, len(currPath))
			copy(copyPath, currPath)
			output = append(output, copyPath)
			return
		}

		if idx >= N || currTotal > target {
			return
		}

		// Take the current element
		currElement := candidates[idx]
		currPath = append(currPath, currElement)
		findCombination(idx, currPath, currTotal+currElement)

		// Skip the current element -- backtrack
		currPath = currPath[:len(currPath)-1]
		findCombination(idx+1, currPath, currTotal)
	}

	findCombination(0, []int{}, 0)

	return output
}

File: 39-combination-sum/combination-sum.cpp
class Solution {
    vector<vector<int>> output;
    vector<int> curr;

    void findCombinations(vector<int>& candidates, int target, int currSum, int idx) {
        if (currSum == target) {
            output.push_back(curr);
            return;
        }

        if (currSum > target || idx >= candidates.size()) {
            return;
        }

        // Take the candidate at current index
        curr.push_back(candidates[idx]);
        currSum += candidates[idx];

        findCombinations(candidates, target, currSum, idx);

        // Skip the candidate at current index
        curr.pop_back();
        currSum -= candidates[idx];

        findCombinations(candidates, target, currSum, idx+1);
    }
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {

        findCombinations(candidates, target, 0, 0);
        
        return output;
    }
};